### ISSUE TITLE
Cannot define patch for a single version

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#1490]
Let's say you're writing a package and only a single version needs to be patched. If you try to define a patch as follows:

``` python
@when('@1.2.2')
def patch(self):
   ....
```

then `spack patch package@1.2.2` works fine. But `spack patch package@1.2.4` crashes with the following error message:

```
==> patch() function failed for py-pyside
==> Error: Package PyPyside does not support patch called with py-pyside@1.2.2.  Options are: py-pyside@1.2.2
```

**Comment by adamjstewart** [comment]
According to directives.py, you should be able to do this:
```python
@directive('patches')
def patch(pkg, url_or_filename, level=1, when=None):
    """Packages can declare patches to apply to source.  You can
       optionally provide a when spec to indicate that a particular
       patch should only be applied when the package's spec meets
       certain conditions (e.g. a particular version).
    """
```

**Comment by alalazo** [comment]
Is somebody working on this (it's in the in progress column of the 0.10.0 project) ? Otherwise I may have time tomorrow to look at the issue.

**Comment by tgamblin** [comment]
I'm going to have a fix for this soon.

**Comment by tgamblin** [comment]
I've got a whole revamp of versions that adds exact versions, but I'm still testing it out today

**Comment by adamjstewart** [comment]
Fixed by #3411
